text_extensions_for_pandas/resources/span_array.js

```diff
@@ -0,0 +1,233 @@
+                        }
+                        entries[i].sets.push(set)
+                    }
+                }
+            }
+
+            return entries
+        }
+
+        constructor(id, begin, end) {
+            this.id = id
+            this.begin = begin
+            this.end = end
+            this.sets = []
+            this.visible = true
+        }
+
+        // Returns only visible sets
+        get valid_sets() {
+            let valid_sets = []
+
+            this.sets.forEach(set => {
+                if(set.entry.visible) valid_sets.push(set)
+            })
+
+            return valid_sets
+        }
+
+        // Returns true if mark should render as a compound set of spans
+        isComplex() {
+            for(let i = 0; i < this.valid_sets.length; i++) {
+                let otherMember = this.valid_sets[i].entry;
+                if(this.valid_sets[i].type == TYPE_OVERLAP && otherMember.visible) {
+                    return true;
+                } else {
+                    if(otherMember.valid_sets.length > 0 && otherMember.visible) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        // Gets the combined span of all connected elements
+        getSetSpan() {
+            let begin = this.begin
+            let end = this.end
+            let highest_id = this.id
+
+            this.valid_sets.forEach(set => {
+                let other = set.entry.getSetSpan()
+                if(other.begin < begin) begin = other.begin
+                if(other.end > end) end = other.end
+                if(other.highest_id > highest_id) highest_id = other.highest_id
+            })
+
+            return {begin: begin, end: end, highest_id: highest_id}
+        }
+    }
+
+    window.SpanArray.Entry = Entry
+
+    // Render DOM
+    function render(doc_text, entries, instance_id, show_offsets) {
+
+        let frag = document.createDocumentFragment()
+        
+        // Render Table
+        if(show_offsets) {
+            let table = document.createElement("table")
+            table.innerHTML = `
+            <thead>
+            <tr>
+                <th></th>
+                <th>begin</th>
+                <th>end</th>
+                <th>context</th>
+            </tr>
+            </thead>`
+            let tbody = document.createElement("tbody")
+            entries.forEach(entry => {
+                let row = document.createElement("tr")
+                row.setAttribute("data-id", entry.id.toString())
+                if(!entry.visible)
+                {
+                    row.classList.add("disabled")
+                }
+
+                row.innerHTML += `
+                <td><b>${entry.id.toString()}</b></td>
+                <td>${entry.begin}</td>
+                <td>${entry.end}</td>
+                <td>${doc_text.substring(entry.begin, entry.end)}</td>`
```

Comment:
Added a call to the ```sanitize()```, expanded the function to include replacing ```$``` with its entity name representation to avoid an accidental LaTeX render. Also added a comment explaining the sanitization.

Reply:
Added comment, sanitization, and additional character replacements to prevent the DOM from accidentally rendering LaTeX.